<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">

    <!-- Bootstrap CSS CDN -->
    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.0/css/bootstrap.min.css" integrity="sha384-9gVQ4dYFwwWSjIDZnLEWnxCjeSWFphJiwGPXr1jddIhOegiu1FwO5qRGvFXOdJZ4" crossorigin="anonymous">
    <!-- CSS -->
    <link rel="stylesheet" href="style.css">
    <!-- Ifremer CSS -->
    <link rel="stylesheet" href="ifremer.css">
    <!-- Scrollbar Custom CSS -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/malihu-custom-scrollbar-plugin/3.1.5/jquery.mCustomScrollbar.min.css">
    <!-- jQuery CDN - Slim version (=without AJAX) -->
    <script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
    <!-- Popper.JS -->
    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.0/umd/popper.min.js" integrity="sha384-cs/chFZiN24E4KMATLdqdvsezGxaGsi4hLGOzlXwp5UZB1LY//20VyM2taTB4QvJ" crossorigin="anonymous"></script>
    <!-- Bootstrap JS -->
    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.0/js/bootstrap.min.js" integrity="sha384-uefMccjFJAIv6A+rW+L4AHf99KvxDjWSu1z9VI8SKNVmz4sk7buKt/6v9KI65qnm" crossorigin="anonymous"></script>
    <!-- jQuery Custom Scroller CDN -->
    <script src="https://cdnjs.cloudflare.com/ajax/libs/malihu-custom-scrollbar-plugin/3.1.5/jquery.mCustomScrollbar.concat.min.js"></script>
    <!--  Font Awesome JS  -->
    <script defer src="https://use.fontawesome.com/releases/v5.0.13/js/solid.js" integrity="sha384-tzzSw1/Vo+0N5UhStP3bvwWPq+uvzCMfrN1fEFe+xBmv1C/AtVX5K0uZtmcHitFZ" crossorigin="anonymous"></script>
    <script defer src="https://use.fontawesome.com/releases/v5.0.13/js/fontawesome.js" integrity="sha384-6OIrr52G08NpOFSZdxxz1xdNSndlD4vdcf/q2myIUVO0VsqaGHJsB0RaBE01VTOY" crossorigin="anonymous"></script>


    <script type="text/javascript">
        $(document).ready(function () {
            $("#sidebar").mCustomScrollbar({
                theme: "minimal"
            });

            $('#sidebarCollapse').on('click', function () {
                $('#sidebar, #content').toggleClass('active');
                $('.collapse.in').toggleClass('in');
                $('a[aria-expanded=true]').attr('aria-expanded', 'false');
            });
        });
    </script>

    <title>SAMBA Report</title>

</head>

<body>
<div>
    <!-- Sidebar  -->
    <nav id="sidebar">
        <div class="sidebar-header" style="text-align: center;">
            <i class="fas fa-home fa-2x"></i><a href="#home" style="font-weight: bold; font-size: 2.0rem"> Home</a>
        </div>
        <div style="text-align: left; padding: 10px">
            Project name : {{ d.projectName }}
        </div>
        <div style="text-align: left; padding: 10px">
            Analysis date : {{ d.run_date }}
        </div>
        <ul class="list-unstyled components">
            <li><a href="#description">Description</a></li>
            {% if d.steps.stats_only == 'false' %}
            <li class="active">
                <a href="#BioinformaticSubmenu" data-toggle="collapse" aria-expanded="false" class="dropdown-toggle">Bioinformatic process</a>
                <ul class="collapse list-unstyled" id="BioinformaticSubmenu">
                    {% if d.steps.longreads == 'false' %}
                        {% if d.steps.data_integrity_enable == 'true' %}
                            <li><a href="#data-integrity">Data integrity</a></li>
                        {% endif %}
                        <li><a href="#importing-raw-data">Importing raw data</a></li>
                        {% if d.steps.cutadapt_enable == 'true' %}
                        <li><a href="#primers-removal">Primers removal</a></li>
                        {% endif %}
                        <li><a href="#qc-feature-table">QC and feature table</a></li>
                        {% if d.steps.dbotu3_enable == 'true' %}
                            <li><a href="#asv-clustering">ASV clustering</a></li>
                        {% endif %}
                        <li><a href="#taxo-assign">Taxonomic assignation</a></li>
                        {% if d.steps.filtering_tax_enable == 'true' %}
                            <li><a href="#tax-filtering">Taxonomy filtering</a></li>
                        {% endif %}
                        {% if d.steps.microDecon_enable == 'true' %}
                            <li><a href="#microdecon">Samples decontamination</a></li>
                        {% endif %}
                        {% if d.steps.picrust2_enable == 'true' %}
                            <li><a href="#picrsut">Functional predictions</a></li>
                        {% endif %}
                        <li><a href="#ancom">Differential abundance</a></li>
                        <li><a href="#final-output">Final outputs</a></li>
                    {% else %}
                        <li><a href="#taxo-assign">Taxonomic assignation</a></li>
                    {% endif %}
                </ul>
            </li>
            {% endif %}
            <li class="active">
                <a href="#StatisticalSubmenu" data-toggle="collapse" aria-expanded="false" class="dropdown-toggle">Statistical analyses</a>
                <ul class="collapse list-unstyled" id="StatisticalSubmenu">
                    {% if d.steps.stats_alpha_enable == 'true' %}
                    <li><a href="#alpha-diversity">Alpha diversity</a></li>
                    {% endif %}
                    {% if d.steps.stats_beta_enable == 'true' %}
                    <li><a href="#beta-diversity">Beta diversity</a></li>
                    {% endif %}
                    {% if d.steps.stats_desc_comp_enable == 'true' %}
                    <li><a href="#decriptive-comparison">Descriptive comparison</a></li>
                    {% endif %}
                </ul>
            </li>
            <li><a href="#about">About</a></li>
            <hr>
            <div style="text-align: center; font-size: 1.0rem; color: #f0f0f0; margin-left: 8px; margin-right: 8px">samba by SeBiMER</div>
            <br>
            <div style="text-align: center; font-size: 1.0rem; color: #f0f0f0"><i class="fas fa-envelope"></i><a href="mailto: samba-sebimer@ifremer.fr"> Contact</a></div>
        </ul>
        </nav>
        <!-- Page Content  -->
    <div id="content">
        <p class="subtitle">SAMBA: Standardized and Automated MetaBarcoding Analyses workflow</p>
        <hr class="title">
        <br>
        <h1 id="description">Description of the workflow</h1>
        <br>
        <div style="text-align: justify">
            The <b>SAMBA workflow</b>, developed by the <b>SeBiMER</b> (Ifremer's Bioinformatics Core Facility) is an open-source modular workflow
            to process eDNA metabarcoding data. SAMBA is developped using the NextFlow workflow manager (Di Tommaso <i>et al.</i>, 2017). SAMBA is built
            around three main parts: data integrity checking, bioinformatics processes and statistical analyses. The SAMBA checking process
            allows to verify the integrity of the raw data. All bioinformatics processes are mainly based on the use of the next-generation
            microbiome bioinformatics platform <b>QIIME 2</b> (Bolyen <i>et al.</i>, 2019 ; version 2020.2) and on the approach of grouping
            sequences in ASV (Amplicon Sequence Variants) using <b>DADA2</b> (Callahan <i>et al.</i>, 2016). It also performs extensives analyses
            of the alpha- and beta-diversity using homemade R scripts (R CORE TEAM, 2020). SAMBA offers a real alternative to the complex use of
            a suite of command line bioinformatics tools while providing access to state-of-the-art methods and tools in the field.
            The SAMBA source code, documentation and installation instructions are freely available at <a class="link" href="https://github.com/ifremer-bioinformatics/samba">SeBiMER GitHub</a>.
        </div>
        <br>
        <p align="center">
            <a class="link" target="_blank" href="samba_wf.png">
                <img src="samba_wf.png" style="width:75%">
            </a>
        </p>
        <br>
        <div style="text-align: justify">
          In order to achieve the complete analysis, SAMBA uses a range of state-of-the-art software and methods which you will find the list below :
        </div>
        <br>
        <table id="tools_list">
            <tr>
                <th>ToolsÂ and softwares</th>
                <th style="text-align: center">Version</th>
            </tr>
            <tr>
                <td><a class="link" target="_blank" href="https://github.com/ifremer-bioinformatics/samba">SAMBA</a></td>
                <td style="text-align: center">{{d.soft.samba}}</td>
            </tr>
            <tr>
                <th colspan="2"><i>Required</i></th>
            </tr>
            <tr>
                <td><a class="link" target="_blank" href="https://www.nextflow.io/index.html">Nextflow</a></td>
                <td style="text-align: center">{{d.soft.nextflow}}</td>
            </tr>
            <tr>
                <th colspan="2"><i>Included in SAMBA</i></th>
            </tr>
            <tr>
                <td>
                    <a class="link" target="_blank" href="https://qiime2.org/">QIIME 2</a>
                    <ul>
                        <li><a class="link" target="_blank" href="https://github.com/cduvallet/q2-dbotu">q2-dbotu</a></li>
                        <li><a class="link" target="_blank" href="https://library.qiime2.org/plugins/q2-picrust2/13/">q2-picrust2</a></li>
                    </ul>
                </td>
                <td style="text-align: center">{{d.soft.qiime2}}</td>
            </tr>
            <tr>
                <td>
                    <a class="link" target="_blank" href="https://cran.r-project.org/">R</a>
                </td>
                <td style="text-align: center">{{d.soft.R}}</td>
            </tr>
            <tr>
                <td><a class="link" target="_blank" href="https://github.com/donaldtmcknight/microDecon">microDecon</a></td>
                <td style="text-align: center">{{d.soft.microdecon}}</td>
            </tr>
            <tr>
                <td><a class="link" target="_blank" href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a></td>
                <td style="text-align: center">{{d.soft.DESeq2}}</td>
            </tr>
            <tr>
                <td><a class="link" target="_blank" href="https://www.bioconductor.org/packages/release/bioc/html/metagenomeSeq.html">metagenomeSeq</a></td>
                <td style="text-align: center">{{d.soft.metagenomeSeq}}</td>
            </tr>
            <tr>
                <td><a class="link" target="_blank" href="https://microbiome.github.io/tutorials/">microbiome</a></td>
                <td style="text-align: center">{{d.soft.microbiome}}</td>
            </tr>
            <tr>
                <td><a class="link" target="_blank" href="https://joey711.github.io/phyloseq/">phyloseq</a></td>
                <td style="text-align: center">{{d.soft.phyloseq}}</td>
            </tr>
            <tr>
                <td><a class="link" target="_blank" href="https://cran.r-project.org/web/packages/vegan/vegan.pdf">vegan</a></td>
                <td style="text-align: center">{{d.soft.vegan}}</td>
            </tr>
            <tr>
                <td><a class="link" target="_blank" href="https://github.com/hms-dbmi/UpSetR">UpSetR</a></td>
                <td style="text-align: center">{{d.soft.UpSetR}}</td>
            </tr>
            <tr>
                <td><a class="link" target="_blank" href="https://github.com/lh3/minimap2">Minimap2</a></td>
                <td style="text-align: center">{{d.soft.minimap2}}</td>
            </tr>
        </table>
        <br>
        {% if d.steps.stats_only == 'false' %}
        <h1 id="bioinformatic">Bioinformatic process</h1>
        {% if d.steps.data_integrity_enable == 'true' %}
        <h3 id ="data-integrity">Data integrity [optional]</h3>
        <br>
        <div style="text-align: justify">
            This first step allows to <b>analyze the integrity of your data</b> in order to identify
            potential problems related to the sequencing processes. It checks:
            <ul>
                <li>that each read is correctly associated with the proper sample (sequence barcode verification)</li>
                <li>that they come from a single sequencer</li>
                <li>the efficiency of forward and reverse PCR amplification</li>
            </ul>
            <b>This step is carried out by a homemade script</b>
        </div>
        <br>
        <div class="results">Results</div>
        <hr class="results">
        <ul style="margin: 0 ; padding: 1 ;"><li>All data integrity verification are summarized in <a class="link" href="./data_integrity.txt">this csv file</a></li></ul>
        <br>
        {% endif %}
        {% if d.steps.longreads == 'false' %}
        <h3 id="importing-raw-data">Importing raw data</h3>
        <br>
        <div style="text-align: justify">
            The step performs the import of sequencing data directly from DATAREF <b>into
            a QIIME 2 specific format</b>. In addition, <b>descriptive statistics of your data are generated
            </b>. The SAMBA workflow ran for you the <a class="link" target="_blank" href="./cmd/q2_import_complete.sh">following commands</a>
        </div>
        <br>
        <div class="results">Results</div>
        <hr class="results">
        <i class="fas fa-folder" style="margin-left: 20px; color: #007bff"></i><a class="link" target="_blank" href="./import_output" style="font-size: 1.4rem"> Output folder</a>
        <br>
        <br>
        <ul style="margin: 0 ; padding: 1 ;">
            <li>Sample repartition according to the sequence count</li>
        </ul>
        <br>
        <p align="center"> <img src="./import_output/demultiplex-summary.png" style="width:40%"></p>
        <ul style="margin: 0 ; padding: 1 ;"><li>All descriptive statistics of your samples are available <a class="link" target="_blank" href="./import_output/overview.html">here</a> (html output)</li></ul>
        <br>
        {% if d.steps.cutadapt_enable == 'true' %}
        <h3 id="primers-removal">Primers removal</h3>
        <br>
        <div style="text-align: justify">
           The third step is to remove the primers using <b>the cutadapt plugin</b> available in
           QIIME 2. In addition, <b>descriptive statistics of your data are generated</b>.
           The <a class="link" target="_blank" href="./cmd/q2_cutadapt_complete.sh">following commands</a> made this step possible using the parameters that you have
           defined in the parameters configuration file :
        </div>
        <br>
        <i class="fas fa-terminal" style="margin-left: 20px; color: #3e3e3e"></i><a style="color: #3e3e3e; font-weight: bold"> User-defined Cutadapt parameters</a>
        <ul style="margin: 0 ; padding: 1 ;">
            {% if d.singleEnd == 'false' %}
            <li class="cmdline">Mode: <span class="highlight">Paired-end</span></li>
            <li class="cmdline">Forward primer sequence: <span class="highlight">{{d.cutadapt.primerF}}</span></li>
            <li class="cmdline">Reverse primer sequence: <span class="highlight">{{d.cutadapt.primerR}}</span></li>
            <li class="cmdline">Overlap: <span class="highlight">{{d.cutadapt.overlap}}</span></li>
            {% else %}
            <li class="cmdline">Mode: <span class="highlight">Single-end</span></li>
            <li class="cmdline">Primer sequence: <span class="highlight">{d.cutadapt.primerF}}</span></li>
            {% endif %}
            <li class="cmdline">Error rate: <span class="highlight">{{d.cutadapt.errorRate}}</span></li>
        </ul>
        <br>
        <div class="results">Results</div>
        <hr class="results">
        <i class="fas fa-folder" style="margin-left: 20px; color: #007bff"></i><a class="link" target="_blank" href="./trimmed_output" style="font-size: 1.4rem"> Output folder</a>
        <br>
        <br>
        <ul style="margin: 0 ; padding: 1 ;">
            <li>Sample repartition according to the sequence count</li>
        </ul>
        <br>
        <p align="center"> <img src="./trimmed_output/demultiplex-summary.png" style="width:40%"><br></p>
        <ul style="margin:0 ; padding: 1 ;">
            <li>All descriptive statistics of your samples after the trimming are available <a class="link" target="_blank" href="./trimmed_output/overview.html">here</a> (html output)</li>
            <li>The quality of your data before any quality filtering step is viewable <a class="link" target="_blank" href="./trimmed_output/quality-plot.html">here</a> (html output)</li>
        </ul>
        <br>
        {% endif %}
        <!--  -->
        <!-- Sequence quality control and feature table construction -->
        <!--  -->
        <h3 id="qc-feature-table">Sequence quality control and feature table construction</h3>
        <br>
        <div style="text-align: justify">
            The following step of the workflow is to filter the quality of the sequences according
            to the parameters defined yourself by having considered the quality of your data in the
            previous step. Also to assemble the forward and reverse sequences, and to identify and remove the
            chimeras. To performed this step, the <b>DADA2 R package</b> was used through QIIME 2
            using the <a class="link" target="_blank" href="./cmd/q2_dada2_complete.sh">following commands</a> and your parameters :
        </div>
        <br>
        <i class="fas fa-terminal" style="margin-left: 20px; color: #3e3e3e"></i><a style="color: #3e3e3e; font-weight: bold"> User-defined DADA2 parameters</a>
        <ul style="margin: 0 ; padding: 1 ;">
            {% if d.singleEnd == 'false' %}
            <li class="cmdline">Mode: <span class="highlight">Paired-end</span></li>
            <li class="cmdline">Number of bases trimmed in 5' of forward reads: <span class="highlight">{{d.dada2.FtrimLeft}}</span></li>
            <li class="cmdline">Number of bases trimmed in 5' of reverse reads: <span class="highlight">{{d.dada2.RtrimLeft}}</span></li>
            <li class="cmdline">Length to trim forward reads (0 for no trimming): <span class="highlight">{{d.dada2.FtruncLen}}</span></li>
            <li class="cmdline">Length to trim reverse reads (0 for no trimming): <span class="highlight">{{d.dada2.RtruncLen}}</span></li>
            <li class="cmdline">Max error rate allowed in forward reads: <span class="highlight">{{d.dada2.FmaxEE}}</span></li>
            <li class="cmdline">Max error rate allowed in reverse reads: <span class="highlight">{{d.dada2.RmaxEE}}</span></li>
            {% else %}
            <li class="cmdline">Mode: <span class="highlight">Single-end</span></li>
            <li class="cmdline">Number of bases trimmed in 5' of reads: <span class="highlight">{{d.dada2.trimLeft}}</span></li>
            <li class="cmdline">Length to trim reads: <span class="highlight">{{d.dada2.FtruncLen}}</span></li>
            <li class="cmdline">Max error rate allowed: <span class="highlight">{{d.dada2.FmaxEE}}</span></li>
            {% endif %}
            <li class="cmdline">Minimal quality score allowed: <span class="highlight">{{d.dada2.minQ}}</span></li>
            <li class="cmdline">Method for chimeras detection: <span class="highlight">{{d.dada2.chimeras}}</span></li>
        </ul>
        <br>
        <div class="results">Results</div>
        <hr class="results">
        <i class="fas fa-folder" style="margin-left: 20px; color: #007bff"></i><a class="link" target="_blank" href="./dada2_output" style="font-size: 1.4rem"> Output folder</a>
        <br>
        <br>
        <ul style="margin: 0 ; padding: 1 ;text-align: justify">
            <li>The dynamics of the different step of filtering can be visualized in this <a class="link" target="_blank" href="./dada2_output/index.html">html file</a> and
                are also available in a <a class="link" target="_blank" href="./dada2_output/metadata.tsv">tabulated file</a></li>
            <li>Distribution of sequences in samples</li>
            <br>
            <p align="center"><img src="./dada2_output/sample-frequencies.png" style="width:40%"></p>
            <br>
            <li>Feature details are available <a class="link" target="_blank" href="./dada2_output/feature-frequency-detail.html">here</a> (html output)</li>
            <li>Details about the samples can be found by going to <a class="link" target="_blank" href="./dada2_output/sample-frequency-detail.html">this interactive html page</a></li>
            <li>Finally, you can retrieved the reference sequences of your ASVs in <a class="link" target="_blank" href="./dada2_output/sequences.fasta">this fasta</a></li>
            <br>
            <font color="red"><b>A total of {{ r.dada2.asv_count }} ASVs was obtained at this step</b></font>
        </ul>
        <br>
        <!--  -->
        <!-- ASV clustering -->
        <!--  -->
        {% if d.steps.dbotu3_enable == 'true' %}
        <h3 id="asv-clustering">ASV clustering [optional]</h3>
        <br>
        <div style="text-align: justify">
            The following step of the workflow is to cluster ASV according to sequence similarity and
            abundance profil. This allow to takes into account the overestimation of diversity
            produced by DADA2. It also reduces possible PCR errors. To performed this step,
            the <b>dbOTU3 algorithm</b> was used through QIIME 2 using the <a class="link" target="_blank" href="./cmd/q2_dbotu3_complete.sh">following commands</a> and the parameters :
        </div>
        <br>
        <i class="fas fa-terminal" style="margin-left: 20px; color: #3e3e3e"></i><a style="color: #3e3e3e; font-weight: bold"> User-defined dbOTU3 parameters</a>
        <ul style="margin: 0 ; padding: 1 ;">
            <li class="cmdline">Genetic criterion: <span class="highlight">{{d.dbotu3.gen_crit}}</span></li>
            <li class="cmdline">Abundance criterion: <span class="highlight">{{d.dbotu3.abund_crit}}</span></li>
            <li class="cmdline">Pvalue criterion: <span class="highlight">{{d.dbotu3.pval_crit}}</span></li>
        </ul>
        <br>
        <div class="results">Results</div>
        <hr class="results">
        <i class="fas fa-folder" style="margin-left: 20px; color: #007bff"></i><a class="link" target="_blank" href="./dbotu3_output" style="font-size: 1.4rem"> Output folder</a>
        <br>
        <br>
        <ul style="margin: 0 ; padding: 1 ;text-align: justify">
            <li>Overview of the results available <a class="link" target="_blank" href="./dbotu3_output/index.html">here</a> (html output)</li>
            <li>Feature details are available <a class="link" target="_blank" href="./dbotu3_output/feature-frequency-detail.html">here</a> (html output)</li>
            <li>Details about the samples can be found by going to <a class="link" target="_blank" href="./dbotu3_output/sample-frequency-detail.html">this interactive html page</a></li>
            <li>Finally, you can retrieved the reference sequences of your ASVs in <a class="link" target="_blank" href="./dbotu3_output/sequences.fasta">this fasta</a></li>
            <br>
            <font color="red"><b>A total of {{ r.dbotu3.asv_count }} ASVs remain after this step. dbOTU3 allowed to cluster {{ r.dbotu3.clustering }} ASVs with others ASVs (i.e. about {{ r.dbotu3.clustering_perc }}% clustering). </b></font>
        </ul>
        <br>
        {% endif %}
        {% endif %}
        <!--  -->
        <!-- Taxonomic assignation -->
        <!--  -->
        <h3 id="taxo-assign">Taxonomic assignation</h3>
        <br>
        {% if d.steps.longreads == 'false' %}
        <div style="text-align: justify">
            The taxonomic assignation performed during the fourth step of the workflow allowed to
            affiliate each ASV to a taxonomy by using as reference the database defined by yourself.<a class="link" target="_blank" href="./cmd/q2_taxonomy_complete.sh">Commands executed</a>
        </div>
        <br>
        <i class="fas fa-terminal" style="margin-left: 20px; color: #3e3e3e"></i><a style="color: #3e3e3e; font-weight: bold"> User-defined QIIME 2 parameters</a>
        <ul style="margin: 0 ; padding: 1 ;">
            <li class="cmdline">Minimal confidence allowed: <span class="highlight">{{d.taxonomy.confidence}}</span></li>
            <li class="cmdline">Taxonomic database used: <span class="highlight">{{d.taxonomy.database}}</span></li>
        </ul>
        <br>
        <div class="results">Results</div>
        <hr class="results">
        <i class="fas fa-folder" style="margin-left: 20px; color: #007bff"></i><a class="link" target="_blank" href="./taxo_output" style="font-size: 1.4rem"> Output folder</a>
        <br>
        <br>
        <ul style="margin: 0 ; padding: 1 ;text-align: justify">
            <li>The result of the taxonomic affiliation is available in <a class="link" target="_blank" href="./taxo_output/index.html">html</a> and
                <a class="link" target="_blank" href="./ASV_taxonomy.tsv">tabulated</a> formats</li>
        </ul>
        {% else  %}
        <div style="text-align: justify">
            The taxonomic assignation allowed to affiliate each <b>read</b> to a taxonomy by using as reference the database defined by yourself.
        </div>
        <br>
        <i class="fas fa-terminal" style="margin-left: 20px; color: #3e3e3e"></i><a style="color: #3e3e3e; font-weight: bold"> User-defined Minimap2 and Taxonomy parameters</a>
        <ul style="margin: 0 ; padding: 1 ;">
            <li class="cmdline">Sequencing technology: <span class="highlight">{{d.lr.lr_type}}</span></li>
            <li class="cmdline">Minimal rank level: <span class="highlight">{{d.lr.lr_rank}}</span></li>
            <li class="cmdline">Reference database: <span class="highlight">{{d.lr.lr_tax_fna}}</span></li>
            <li class="cmdline">Reference taxonomy: <span class="highlight">{{d.lr.lr_tax_flat}}</span></li>
        </ul>
        <br>
        <div class="results">Results</div>
        <hr class="results">
        <i class="fas fa-folder" style="margin-left: 20px; color: #007bff"></i><a class="link" target="_blank" href="./taxo_output" style="font-size: 1.4rem"> Output folder</a>
        <br>
        <br>
        <ul style="margin: 0 ; padding: 1 ;text-align: justify">
            <li>The result of the taxonomic affiliation is available in <a class="link" target="_blank" href="./02_lr_mapping/samples.tsv">tabulated</a> formats</li>
        </ul>
        {% endif %}
        <br>
        <!--  -->
        <!-- Taxonomy filtering [optional] -->
        <!--  -->
        {% if d.steps.filtering_tax_enable == 'true' %}
        <h3 id="tax-filtering">Taxonomy filtering [optional]</h3>
        <br>
        <div style="text-align: justify">
            This step allows you to filter the ASV table and ASV sequences based on the taxonomy assignation and remove user-defined taxa.
        </div>
        <br>
        <i class="fas fa-terminal" style="margin-left: 20px; color: #3e3e3e"></i><a style="color: #3e3e3e; font-weight: bold"> User-defined 'Taxonomy filtering' parameters</a>
        <ul style="margin: 0 ; padding: 1 ;">
            <li class="cmdline">List of taxa to exclude: <span class="highlight">{{d.tax_filtering.tax_to_exclude}}</span></li>
        </ul>
        <br>
        <div class="results">Results</div>
        <hr class="results">
        <i class="fas fa-folder" style="margin-left: 20px; color: #007bff"></i><a class="link" target="_blank" href="./tax_filtering" style="font-size: 1.4rem"> Output folder</a>
        <br>
        <br>
        <ul style="margin: 0 ; padding: 1 ;text-align: justify">
            <li>ASV table: <a class="link" target="_blank" href="./tax_filtering/tax_filtered_table_with_tax.tsv">taxonomy filtered ASV table</a></li>
            <li>Representative sequence: <a class="link" target="_blank" href="./tax_filtering/tax_filtered_output/sequences.fasta">taxonomy filtered ASV sequences</a></li>
            <li>Descriptive statistics after the filtering step: <a class="link" target="_blank" href="./tax_filtering/tax_filtered_output/index.html">html file</a></li>
            <br>
            <font color="red"><b>A total of {{ r.tax_filtering.asv_rm }} ASVs, assigned to taxa to exclude, were removed. It remains {{ r.tax_filtering.asv_count }} ASVs.</b></font>
        </ul>
        <br>
        {% endif %}
        <!--  -->
        <!-- Samples decontamination (microDecon) [optional] -->
        <!--  -->
        {% if d.steps.microDecon_enable == 'true' %}
        <h3 id="microdecon">Samples decontamination (microDecon) [optional]</h3>
        <br>
        <div style="text-align: justify">
            If you have sequenced control samples, this step allows you to decontaminate your samples from the sequences identified in these controls. This step is carried out using microDecon and a homemade R script.
        </div>
        <br>
        <i class="fas fa-terminal" style="margin-left: 20px; color: #3e3e3e"></i><a style="color: #3e3e3e; font-weight: bold"> User-defined microDecon parameters</a>
        <ul style="margin: 0 ; padding: 1 ;">
            <li class="cmdline">Control list: <span class="highlight">{{d.microdecon.control_list}}</span></li>
            <li class="cmdline">Number of blanks: <span class="highlight">{{d.microdecon.nb_controls}}</span></li>
            <li class="cmdline">Number of  samples: <span class="highlight">{{d.microdecon.nb_samples}}</span></li>
        </ul>
        <br>
        <div class="results">Results</div>
        <hr class="results">
        <i class="fas fa-folder" style="margin-left: 20px; color: #007bff"></i><a class="link" target="_blank" href="./microDecon" style="font-size: 1.4rem"> Output folder</a>
        <br>
        <br>
        <ul style="margin: 0 ; padding: 1 ;text-align: justify">
            <li>New ASV table: <a class="link" target="_blank" href="./microDecon/decontaminated_ASV_table.tsv">decontaminated table</a></li>
            <li>Representative sequence: <a class="link" target="_blank" href="./microDecon/decontaminated_ASV.fasta">decontaminated ASV sequence</a></li>
            <li>All ASVs listed <a class="link" target="_blank" href="./microDecon/ASV_removed.txt">here</a> were removed because they were in the majority in the control samples         </li>
            <li>For all ASVs listed <a class="link" target="_blank" href="./microDecon/abundance_removed.txt">here</a>, a proportion of sequences were removed in each sample. Note that this file also contains the ASVs unique to the controls which have therefore been removed.</li>
            <br>
            <font color="red"><b>A total of {{ r.microdecon.asv_count }} ASVs remain after this step. MicroDecon removed {{ r.microdecon.asv_rm }} ASVs ({{ r.microdecon.asv_rm_ctrl }} ASVs were unique to controls).</b></font>
        </ul>
        <br>
        {% endif %}
        <!--  -->
        <!-- Functional predictions (PICRUSt2) [optional] -->
        <!--  -->
        {% if d.steps.picrust2_enable == 'true' %}
        <h3 id="picrsut">Functional predictions (PICRUSt2) [optional]</h3>
        <br>
        <div style="text-align: justify">
            This optional step allows you to predic the functional potential of the communities using PICRUSt2
        </div>
        <br>
        <i class="fas fa-terminal" style="margin-left: 20px; color: #3e3e3e"></i><a style="color: #3e3e3e; font-weight: bold"> User-defined PICRUSt2 parameters</a>
        <ul style="margin: 0 ; padding: 1 ;">
            <li class="cmdline">HSP method: <span class="highlight">{{d.picrust2.method}}</span></li>
            <li class="cmdline">Max nsti: <span class="highlight">{{d.picrust2.nsti}}</span></li>
        </ul>
        <br>
        <div class="results">Results</div>
        <hr class="results">
        <i class="fas fa-folder" style="margin-left: 20px; color: #007bff"></i><a class="link" target="_blank" href="./picrust2_output" style="font-size: 1.4rem"> Output folder</a>
        <br>
        <br>
        <ul style="margin: 0 ; padding: 1 ;text-align: justify">
            <!--  EC predictions  -->
            <li>EC predictions: <a class="link" target="_blank" href="./picrust2_output/q2-picrust2_output/ec_metagenome_exported/ec_metagenome_predictions_with-descriptions.tsv">EC metagenome predictions</a></li>
            <div class="row">
                {% for ec in r.picrust.var_tested %}
                <div class="column" style="text-align: center;">
                    <span class="highlight">{{ec}}</span>
                    <iframe src="picrust2_output/EC_functional_predictions_NMDS_{{ec}}_interactive.html" style="width:100% ; height:100%"></iframe>
                </div>
                {% endfor %}
            </div>
            <!--  KO predictions  -->
            <li>KO predictions: <a class="link" target="_blank" href="./picrust2_output/q2-picrust2_output/ko_metagenome_exported/ko_metagenome_predictions_with-descriptions.tsv">KO metagenome predictions</a></li>
            <div class="row">
                {% for ko in r.picrust.var_tested %}
                <div class="column" style="text-align: center;">
                    <span class="highlight">{{ko}}</span>
                    <iframe src="picrust2_output/KO_functional_predictions_NMDS_{{ko}}_interactive.html" style="width:100% ; height:100%"></iframe>
                </div>
                {% endfor %}
            </div>
            <!--  METACYC predictions  -->
            <li>METACYC predictions: <a class="link" target="_blank" href="./picrust2_output/q2-picrust2_output/pathway_abundance_exported/pathway_abundance_predictions_with-descriptions.tsv">METACYC abundance predictions</a></li>
            <div class="row">
                {% for metacyc in r.picrust.var_tested %}
                <div class="column" style="text-align: center;">
                    <span class="highlight">{{metacyc}}</span>
                    <iframe src="picrust2_output/MetaCyc_functional_predictions_NMDS_{{metacyc}}_interactive.html" style="width:100% ; height:100%"></iframe>
                </div>
                {% endfor %}
            </div>
        </ul>
        <br>
        {% endif %}
        <!--  end picrust  -->
        <!--  -->
        <!--  Differential abundance testing using ANCOM -->
        <!--  -->
        {% if d.steps.ancom_enable == 'true' %}
        <h3 id="ancom">Differential abundance testing using ANCOM</h3>
        <br>
        <div style="text-align: justify">
            This step allows you to test if there are differentially abundant ASVs depending on the variable of interest that you have specified using the <a class="link" href="./cmd/q2_ancom_complete.sh">following commands</a> :</span>
        </div>
        <br>
        <div class="results">Results</div>
        <hr class="results">
        <i class="fas fa-folder" style="margin-left: 20px; color: #007bff"></i><a class="link" target="_blank" href="./ancom_output" style="font-size: 1.4rem"> Output folder</a>
        <br>
        <br>
        <div style="text-align: justify">
            List of all tested variables :
        </div>
        <ul>
            {% for v in r.ancom.var %}
            <li><span class="highlight">{{ v }}</span></li>
            {% endfor %}
        </ul>
        <br>
        {% for v in r.ancom.var %}
        <div style="text-align: justify">
            <b>ANCOM analysis based on the <span class="highlight">{{ v }}</span> variable :</b>
        </div>
        <ul>
            - The full analysis is available in the following directories:
            <a class="link" target="_blank" href="./ancom_output/export_ancom_{{v}}">ASV level</a> ;
            <a class="link" target="_blank" href="./ancom_output/export_ancom_{{v}}_family">family level</a> ;
            <a class="link" target="_blank" href="./ancom_output/export_ancom_{{v}}_genus">genus level</a>
            <br>
            - The reports generated by QIIME 2 are available by following the following links:
            <a class="link" target="_blank" href="./ancom_output/export_ancom_{{d.stats.ancom_var}}/index.html">ASV level</a> ;
            <a class="link" target="_blank" href="./ancom_output/export_ancom_{{d.stats.ancom_var}}_family/index.html">family level</a> ;
            <a class="link" target="_blank" href="./ancom_output/export_ancom_{{d.stats.ancom_var}}_genus/index.html">genus level</a>
            <br>
            {% if r.ancom[v].base > 0 %}
            - Number of significantly differentially abundant ASVs: <span class="highlight">{{r.ancom[v].base}}</span> <a class="link" target="_blank" href="./ancom_output/export_ancom_{{v}}/ancom.tsv">[Full list]</a>
            {% else %}             
            - No significantly differentially abundant ASVs
            {% endif %}
            <br>
            {% if r.ancom[v].family > 0 %}
            - Number of significantly differentially abundant family: <span class="highlight">{{r.ancom[v].family}}</span> <a class="link" target="_blank" href="./ancom_output/export_ancom_{{v}}_family/ancom.tsv">[Full list]</a>
            {% else %}             
            - No significantly differentially abundant family
            {% endif %}
            <br>
            {% if r.ancom[v].genus > 0 %}
            - Number of significantly differentially abundant genus: <span class="highlight"> {{r.ancom[v].genus}}</span> <a class="link" target="_blank" href="./ancom_output/export_ancom_{{v}}_genus/ancom.tsv">[Full list]</a>
            {% else %}             
            - No significantly differentially abundant genus
            {% endif %}
        </ul>
        {% endfor %}
        <!--  end  -->
        <br>
        <!--  end ancom  -->
        <!--  -->
        <!--  Final outputs -->
        <!--  -->
        {% endif %}
        {% if d.steps.longreads == 'false' %}
        <h3 id="final-output">Final outputs</h3>
        <br>
        <div style="text-align: justify">
            This is the last step of the bioinformatic process where the goal is to merge the ASV abundance table with the taxonomy file
        </div>
        <br>
        <div class="results">Results</div>
        <hr class="results">
        <ul style="margin: 0 ; padding: 1 ;text-align: justify">
            <li>The final ASV table can be viewed <a class="link" target="_blank" href="./ASV_table_with_taxonomy.tsv">here</a>.
                A biom file is available <a class="link" target="_blank" href="./ASV_table_with_taxonomy.biom">here</a> formats</li>
        </ul>
        <br>
        {% endif %}
        {% endif %}
        <!--  -->
        <!--  General statistical analyses -->
        <!--  -->
        <h1 id="statistical-analyses">General statistical analyses</h1>
        {% if d.steps.stats_alpha_enable == 'true' %}
        <h3 id="alpha-diversity">Alpha diversity [optional]</h3>
        <br>
        <li>Tested variables :</li>
        <ul>
            {% for var in r.alpha_div.var_tested %}
            <li><span class="highlight">{{ var }}</span></li>
            {% endfor %}
        </ul>
        <li>Diversity index used : "Observed richness / Chao1 / Shannon / InvSimpson / Pielou"</li>
        <!--  -->
        <!-- diversity indices -->
        <!--  -->
        <br>
        <h4>Diversity indices</h4>
        <br>
        {% for var in r.alpha_div.var_tested %}
        <p class="highlight" >{{var}}</p>
        <p align="center">
            <iframe src="R/FIGURES/alpha_diversity/diversity_index/alpha_div_plots_{{var}}_interactive.html" style="width:80% ; height: 900px"></iframe>
        </p>
        {% endfor %}
        <h4>Rarefaction curve</h4>
        <br>
        <p align="center">
            <iframe src="R/FIGURES/alpha_diversity/rarefaction_curve_interactive.html" style="width:80% ; height: 700px"></iframe>
        </p>
        <h4>Taxonomic diversity</h4>
        <br>
        {% if d.steps.longreads == 'false' %}
            {% for var in r.alpha_div.var_tested %}
            <ul>
                <li>Taxonomic barplots group by <span class="highlight">{{var}} :</span></li><br>
                <a class="link" target="_blank" href="R/FIGURES/alpha_diversity/diversity_barplots/{{var}}/barplot_phylum_{{var}}_interactive.html">Barplot at the phylum level</a> ;
                <a class="link" target="_blank" href="R/FIGURES/alpha_diversity/diversity_barplots/{{var}}/barplot_class_{{var}}_interactive.html">Barplot at the class level</a> ;
                <a class="link" target="_blank" href="R/FIGURES/alpha_diversity/diversity_barplots/{{var}}/barplot_order_{{var}}_interactive.html">Barplot at the order level</a> ;
                <a class="link" target="_blank" href="R/FIGURES/alpha_diversity/diversity_barplots/{{var}}/barplot_family_{{var}}_interactive.html">Barplot at the family level</a>
                <br>
            </ul>
            <p align="center">
                <iframe src="R/FIGURES/alpha_diversity/diversity_barplots/{{var}}/barplot_genus_{{var}}_interactive.html" width="100%" height="800"></iframe>
                <br>
            </p>
            {% endfor %}
        {% else %}
            {% for var in r.alpha_div.var_tested %}
                <p align="center">
                    <iframe src="R/FIGURES/alpha_diversity/diversity_barplots/{{var}}/barplot_{{d.lr.lr_rank}}_{{var}}_interactive.html" width="100%" height="800"></iframe>
                    <br>
                </p>
            {% endfor %}
        {% endif %}
        <ul style="margin: 0 ; padding: 1 ;text-align: justify">
            <li>The results of the significance tests carried out for each variable on each diversity indexes can be viewed <a class="link" target="_blank" href="./R/FIGURES/alpha_diversity/index_significance_tests.txt">here</a></li>
        </ul>
        <br>
        {% endif %}
        <!--  end alpha div  -->
        <!--  -->
        <!--  beta div -->
        <!--  -->
        {% if d.steps.stats_beta_enable == 'true' %}

        <h3 id="beta-diversity-1">Beta diversity by index [optional]</h3>
        <br>
        <li>Normalization method : "
            <a class="link" target="_blank" href="./R/FIGURES/beta_diversity_non_normalized">None</a> /
            <a class="link" target="_blank" href="./R/FIGURES/beta_diversity_rarefied">Rarefaction</a>  /
            <a class="link" target="_blank" href="./R/FIGURES/beta_diversity_DESeq2">DESeq2</a> /
            <a class="link" target="_blank" href="./R/FIGURES/beta_diversity_CSS">CSS</a>"
        </li>
        <li>Tested variables :</li>
        <ul>
            {% for v in r.beta_div.var_tested %}
            <li><span class="highlight">{{ v }}</span></li>
            {% endfor %}
        </ul>
        {% if d.steps.longreads == 'false' %}
        <li>Distance matrix : "Jaccard / Bray-Curtis / UniFrac / Weighted UniFrac"</li>
        {% else %}
        <li>Distance matrix : "Jaccard / Bray-Curtis"</li>
        {% endif %}
        <br>

        <!-- by variable -->
        <h3 id="beta-diversity">Beta diversity by variable [optional]</h3>
        <br>
        {% for var in r.beta_div.var_tested %}
        <h4>Ordination and hierarchical clustering for variable: <span class="highlight">{{var}}</span></h4>
        <br>
            {% for method in d.methods %}
        <p><button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#{{var}}{{method}}" aria-expanded="false" aria-controls="collapseExample">Show {{method}} method</button></p>
        <div class="collapse" id="{{var}}{{method}}">
                {% for index in d.indices %}
            <br>
            <p style="margin-left: 20px"><button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#{{var}}{{method}}{{index}}" aria-expanded="false" aria-controls="collapseExample">Show {{index}} plots</button></p>
            <div class="collapse" id="{{var}}{{method}}{{index}}">
                <div style="text-align: center; color: #007bff; font-size: 1.0rem">{{index}}</div>
                <hr class="results">
                <br>
                <div class="row">
                    {% for stat in d.stats %}
                    <div class="column">
                    {% if index != "hclustering" %}
                    <iframe src="R/FIGURES/beta_diversity_{{method}}/{{index}}/{{index}}_{{method}}_{{var}}_{{stat}}_interactive.html" style="width:100% ; height: 100%"></iframe>
                    {% else %}
                    <img src="R/FIGURES/beta_diversity_{{method}}/{{index}}/{{index}}_{{method}}_{{var}}_{{stat}}.png" style="width:100%"></img>
                    {% endif %}
                    </div>
                    {% endfor %}
                </div>
                <hr class="results">
            </div>
            <br>
                {% endfor %}
        </div>
            {% endfor %}
        {% endfor %}
        <h4>Explained variance</h4>
        <br>
        {% for method in d.methods %}
        <p><button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#ExpVar{{method}}_3" aria-expanded="false" aria-controls="collapseExample">Show {{method}} plots</button></p>
        <div class="collapse" id="ExpVar{{method}}_3">
            <div class="row">
                {% for stat in d.stats %}
                <div class="column">
                <iframe src="./R/FIGURES/beta_diversity_{{method}}/ExpVar/pie_ExpVar_{{method}}_{{stat}}_interactive.html" style="width:100% ; height: 100%"></iframe>
                    </a>
                </div>
                {% endfor %}
            </div>
        </div>
        <br>
        {% endfor %}
        {% endif %}
        <!--  end beta div  -->
        <!--  -->
        <!--  desc compar -->
        <!--  -->
        {% if d.steps.stats_desc_comp_enable == 'true' %}
        <h3 id="decriptive-comparison">Descriptive comparison [optional]</h3>
        <br>
        <li>Tested variables :</li>
        <ul>
            {% for var in r.descriptive_comparison.var_tested %}
            <li><span class="highlight">{{ var }}</span></li>
            {% endfor %}
        </ul>
        <br>
        {% for var in r.descriptive_comparison.var_tested %}
        <li>Repartition of ASVs in the samples grouped by <span class="highlight">{{var}}</span>:</li>
        <br>
        <p align="center">
            <a class="link" target="_blank" href="R/FIGURES/descriptive_comparison/upset_plot_{{var}}.png">
                <img src="R/FIGURES/descriptive_comparison/upset_plot_{{var}}.png" style="width:75%">
            </a>
        </p>
        <br>
        {% endfor %}
        <br>
        {% endif %}
        <!--  -->
        <!--  About -->
        <!--  -->
        <h2 id="about">About</h2>
        <h5 style="margin-left: 10px">Contributors</h5>
        <hr class="results">
        <div style="text-align: justify">
            SAMBA is developped by <a style="color: #007bff" href="https://wwz.ifremer.fr/en" target="_blank">Ifremer</a>'s Bioinformatics Core Facility (<a style="color: #007bff" href="https://ifremer-bioinformatics.github.io/" target="_blank">SeBiMER</a>)
            (French Research Institute for Exploitation of the Sea).
        </div>
        <br>
        <h5 style="margin-left: 10px">Citations</h5>
        <hr class="results">
        <div style="text-align: justify">You can cite the NextFlow publication as follows:</div>
        <br>
        <div style="text-align: justify; margin-left: 80px; margin-right: 80px">
          Paolo Di Tommaso, Maria Chatzou, Evan W. Floden, Pablo Prieto Barja, Emilio Palumbo & Cedric Notredame. (2017).
          <a class="link" target="_blank" href="https://www.nature.com/articles/nbt.3820/">Nextflow enables reproducible computational workflows.</a>
          <i>Nature biotechnology</i>. <b>35</b>(4), 316-319.
        </div>
        <br>
        <div style="text-align: justify">You can cite each tool used in SAMBA as follows:</div>
        <br>
        <div style="text-align: justify; margin-left: 80px; margin-right: 80px">
          Benjamin J. Callahan, Paul J. McMurdie, Michael J. Rosen, Andrew W. Han, Amy Jo A. Johnson & Susan P. Holmes. (2016).
          <a class="link" target="_blank" href="https://www.nature.com/articles/nmeth.3869">DADA2: high-resolution sample inference from Illumina amplicon data</a>
          <i>Nature methods</i>. <b>13</b>(7), 581.
        </div>
        <br>
        <div style="text-align: justify; margin-left: 80px; margin-right: 80px">
          Donald T. McKnight, Roger Huerlimann, Deborah S. Bower, Lin Schwarzkopf, Ross A. Alford & Kyall R. Zenger. (2019).
          <a class="link" target="_blank" href="https://onlinelibrary.wiley.com/doi/full/10.1002/edn3.11">microDecon: A highly accurate readâsubtraction tool
          for the postâsequencing removal of contamination in metabarcoding studies.</a> <i>Environmental DNA</i>. <b>1</b>(1), 14-25.
        </div>
        <br>
        <div style="text-align: justify; margin-left: 80px; margin-right: 80px">
          Evan Bolyen, Jai Ram Rideout, Matthew R. Dillon, Nicholas A. Bokulich, Christian C. Abnet, Gabriel A. Al-Ghalith, Harriet Alexander, Eric J. Alm,
          Manimozhiyan Arumugam, Francesco Asnicar, Yang Bai, Jordan E. Bisanz, Kyle Bittinger, Asker Brejnrod, Colin J. Brislawn, C. Titus Brown, Benjamin J. Callahan,
          AndrÃ©s Mauricio Caraballo-RodrÃ­guez, John Chase, Emily K. Cope, Ricardo Da Silva, Christian Diener, Pieter C. Dorrestein, Gavin M. Douglas, Daniel M. Durall,
          Claire Duvallet, Christian F. Edwardson, Madeleine Ernst, Mehrbod Estaki, Jennifer Fouquier, Julia M. Gauglitz, Sean M. Gibbons, Deanna L. Gibson, Antonio Gonzalez,
          Kestrel Gorlick, Jiarong Guo, Benjamin Hillmann, Susan Holmes, Hannes Holste, Curtis Huttenhower, Gavin A. Huttley, Stefan Janssen, Alan K. Jarmusch, Lingjing Jiang,
          Benjamin D. Kaehler, Kyo Bin Kang, Christopher R. Keefe, Paul Keim, Scott T. Kelley, Dan Knights, Irina Koester, Tomasz Kosciolek, Jorden Kreps, Morgan G. I. Langille,
          Joslynn Lee, Ruth Ley, Yong-Xin Liu, Erikka Loftfield, Catherine Lozupone, Massoud Maher, Clarisse Marotz, Bryan D. Martin, Daniel McDonald, Lauren J. McIver,
          Alexey V. Melnik, Jessica L. Metcalf, Sydney C. Morgan, Jamie T. Morton, Ahmad Turan Naimey, Jose A. Navas-Molina, Louis Felix Nothias, Stephanie B. Orchanian,
          Talima Pearson, Samuel L. Peoples, Daniel Petras, Mary Lai Preuss, Elmar Pruesse, Lasse Buur Rasmussen, Adam Rivers, Michael S. Robeson II, Patrick Rosenthal, Nicola Segata,
          Michael Shaffer, Arron Shiffer, Rashmi Sinha, Se Jin Song, John R. Spear, Austin D. Swafford, Luke R. Thompson, Pedro J. Torres, Pauline Trinh, Anupriya Tripathi,
          Peter J. Turnbaugh, Sabah Ul-Hasan, Justin J. J. van der Hooft, Fernando Vargas, Yoshiki VÃ¡zquez-Baeza, Emily Vogtmann, Max von Hippel, William Walters, Yunhu Wan,
          Mingxun Wang, Jonathan Warren, Kyle C. Weber, Charles H. D. Williamson, Amy D. Willis, Zhenjiang Zech Xu, Jesse R. Zaneveld, Yilong Zhang, Qiyun Zhu, Rob Knight
          & J. Gregory Caporaso. (2019). <a class="link" target="_blank" href="https://www.nature.com/articles/s41587-019-0209-9">Reproducible, interactive, scalable and
          extensible microbiome data science using QIIME 2.</a> <i>Nature biotechnology</i>. <b>37</b>(8), 852-857.
        </div>
        <br>
        <div style="text-align: justify; margin-left: 80px; margin-right: 80px">
          Gavin M. Douglas, Vincent J. Maffei, Jesse Zaneveld, Svetlana N. Yurgel, James R. Brown, Christopher M. Taylor, Curtis Huttenhower & Morgan G. I. Langille. (2019).
          <a class="link" target="_blank" href="https://www.biorxiv.org/content/10.1101/672295v1.abstract">PICRUSt2: An improved and extensible approach for metagenome inference.</a>
          <i>BioRxiv</i>. 672295.
        </div>
        <br>
        <div style="text-align: justify; margin-left: 80px; margin-right: 80px">
          Marcel Martin. (2011). <a class="link" target="_blank" href="http://journal.embnet.org/index.php/embnetjournal/article/view/200">Cutadapt removes adapter
          sequences from high-throughput sequencing reads.</a> <i>EMBnet. journal</i>. <b>17</b>(1), 10-12.
        </div>
        <br>
        <div style="text-align: justify; margin-left: 80px; margin-right: 80px">
          R Core Team. (2020). <a class="link" target="_blank" href="https://www.R-project.org/">R: A language and environment for statistical computing.</a>
          R Foundation for Statistical Computing, Vienna, Austria.
        </div>
        <br>
        <div style="text-align: justify; margin-left: 80px; margin-right: 80px">
          Scott W. Olesen, Claire Duvallet & Eric J. Alm. (2017). <a class="link" target="_blank" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5417438/">dbOTU3:
          A new implementation of distribution-based OTU calling.</a> <i>PloS one</i>. <b>12</b>(5).
        </div>
        <br>
        <div style="text-align: justify; margin-left: 80px; margin-right: 80px">
          Siddhartha Mandal, Will Van Treuren, Richard A. White, Merete EggesbÃ¸, Rob Knight & Shyamal D. Peddada. (2015).
          <a class="link" target="_blank" href="https://www.tandfonline.com/doi/full/10.3402/mehd.v26.27663">Analysis of composition of microbiomes: a novel method for studying microbial composition.</a>
          <i>Microbial ecology in health and disease</i>, <b>26</b>(1), 27663.
        </div>
        <br>
    </div>

</div>
</body>
</html>
